File: Sorting Algorithm/MergeSort.cpp
#include<iostream>
using namespace std;
void merge(int arr[], int l, int m, int n){
	int i,j,k;
	int n1 = m-l+1;
	int n2 = n-m;
	int L[n1],R[n2];
	for(int i = 0; i<n1; i++)
		L[i] = arr[l+i];
	for(int j = 0; j<n2; j++)
		R[j] = arr[m+1+j];
	i = 0;
	j = 0;
	k = l;
	while(i<n1 && j<n2){
		if(L[i]>=R[j]){
			arr[k++]=R[j++];
			
		}
		else{
			arr[k++] = L[i++];
		}
	}
	
	while(i<n1){
		arr[k++] = L[i++];
	}
	while(j<n2){
		arr[k++] = R[j++];
	}
}

void MergeSort(int arr[], int l, int r){
	if(l<r){
		int mid = (l+r)/2;
		MergeSort(arr,l,mid);
		MergeSort(arr,mid+1,r);
		merge(arr,l,mid,r);
	}
}

int main(){
	int arr[]  = {4,2,52,6,44,99,2};
	int n = sizeof(arr)/sizeof(arr[0]);
    MergeSort(arr,0,n);
    for(int i = 0; i<n; i++){
    	cout<<arr[i]<<endl;
	}
}

File: datastructure using cpp/merge sort/mergesort.py
#2 way merging arrays

def merge(a1,a2,arr):
    
    i = 0
    j = 0
    k = 0
    m = len(a1)
    n = len(a2)

    while(i<m and j<n):
        if a1[i]>a2[j]:
            arr[k]=a2[j]
            j += 1
            k += 1
        else:
            arr[k] = a1[i]
            i += 1
            k += 1

    while(i<m):
        arr[k] = a1[i]
        i += 1
        k += 1
    while(j<n):
        arr[k] = a2[j]
        j += 1
        k += 1


def mergeSort(arr):
    l = []
    h = []
    if l<h:
        mid = len(arr)//2
        l = arr[:mid]
        h = arr[mid:]
        mergeSort(l)
        mergeSort(h)
        merge(l,h, arr)

        
arr = [19,38,69,20,67]
mergeSort(arr)
print(arr)
            
